hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingContextUnorderedP.java

```diff
@@ -0,0 +1,374 @@
+ */
+
+package com.hazelcast.jet.impl.processor;
+
+import com.hazelcast.internal.util.concurrent.ManyToOneConcurrentArrayQueue;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.Traverser;
+import com.hazelcast.jet.Traversers;
+import com.hazelcast.jet.core.AbstractProcessor;
+import com.hazelcast.jet.core.BroadcastKey;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.ResettableSingletonTraverser;
+import com.hazelcast.jet.core.Watermark;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.jet.datamodel.Tuple3;
+import com.hazelcast.jet.function.DistributedBiFunction;
+import com.hazelcast.jet.function.DistributedFunction;
+import com.hazelcast.jet.impl.util.LoggingUtil;
+import com.hazelcast.jet.pipeline.ContextFactory;
+
+import javax.annotation.CheckReturnValue;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static com.hazelcast.jet.Traversers.traverseIterable;
+import static com.hazelcast.jet.Util.entry;
+import static com.hazelcast.jet.core.BroadcastKey.broadcastKey;
+import static com.hazelcast.jet.datamodel.Tuple2.tuple2;
+import static com.hazelcast.jet.datamodel.Tuple3.tuple3;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.withTryCatch;
+import static com.hazelcast.jet.impl.util.LoggingUtil.logFine;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Processor which, for each received item, emits all the items from the
+ * traverser returned by the given async item-to-traverser function, using a
+ * context object.
+ * <p>
+ * This processors might reorder items: results are emitted as they are
+ * asynchronously delivered. However, this processor doesn't reorder items with
+ * respect to the watermarks that followed them. That is, a watermark is
+ * guaranteed to be emitted <i>after</i> results for all items that occurred
+ * before it are emitted.
+ *
+ * @param <C> context object type
+ * @param <T> received item type
+ * @param <K> extracted key type
+ * @param <R> emitted item type
+ */
+public final class AsyncTransformUsingContextUnorderedP<C, T, K, R> extends AbstractProcessor {
+
+    private final ContextFactory<C> contextFactory;
+    private final DistributedBiFunction<? super C, ? super T, CompletableFuture<Traverser<R>>> callAsyncFn;
+    private final Function<? super T, ? extends K> extractKeyFn;
+
+    private C contextObject;
+    private ManyToOneConcurrentArrayQueue<Tuple3<T, Long, Object>> resultQueue;
+    // TODO we can use more efficient structure: we only remove from the beginning and add to the end
```

Comment:
I planned to leave it like this. The map will always contain at most 10 items or so, not sure if it's worth added complexity. The number is equal to number of postponed WMs, that is it will roughly be `maxLatencyOfAsyncOp/watermarkFrequence`.